### PR TITLE
Paper: Pandera: Going Beyond Pandas Data Validation

### PR DESCRIPTION
**Pull request by cosmicBboy**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

*Editor*: @cbcunc

*Reviewers*: @hongsupshin, @camerse

### CODE REVIEW

**Comment by deniederhut** [comment]
@scoobies assign @cbcunc as editor

**Comment by scoobies** [comment]
Assigned! @cbcunc is now the editor

**Comment by hongsupshin** [comment]
@cosmicBboy Hi Niels, I will be one of the reviewers of your manuscript :) Will start adding comments soon.

**Comment by deniederhut** [comment]
@scoobies add @hongsupshin to reviewers

**Comment by scoobies** [comment]
@hongsupshin added to the reviewers list!

**Comment by deniederhut** [comment]
@scoobies code of conduct

**Comment by scoobies** [comment]
Our CoC: https://www.scipy2023.scipy.org/code-of-conduct

**Comment by scoobies** [comment]
Reports of abusive or harassing behavior may be reported to scipy@enthought.com

**Comment by deniederhut** [comment]
@scoobies add @camerse to reviewers

**Comment by scoobies** [comment]
@camerse added to the reviewers list!

**Comment by cbcunc** [comment]
Editor says paper is complete and builds. Awaiting review.

File: papers/niels_bantilan/paper.rst [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+
+Origins
+-------
+
+Pandera started as a small project in 2018 with the goal of providing a lightweight,
+flexible, and expressive API to validate pandas DataFrames :cite:`mckinney-proc-scipy-2010`.
+This introductory section provides a brief primer on data validation with pandera,
+providing insights into how its design facilitates code-first schema authoring
+and maintenance. The operating assumption is that this, in turn, gives rise to
+safer and more robust data pipelines.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:52]: Consider moving this after the "why validate data". I think that way readers might be better motivated. Besides, here we say "validate pandas DataFrames" but the definition of validation is not explained yet.

File: papers/niels_bantilan/paper.rst [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+
+Why Validate Data?
+++++++++++++++++++
+
+Data validation is the act of falsifying data against a particular set of
+assumptions, expressed as a schema of validation rules. These rules are explicitly
+established by the analyst without interference from automated processes, like
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: Personally I prefer the term "ML practitioner" better than "analyst" because then we imply that the importance of data validation should be recognized by everyone in ML research and production pipeline. Check out this paper:

https://storage.googleapis.com/pub-tools-public-publication-data/pdf/0d556e45afc54afeb2eb6b51a9bc1827b9961ff4.pdf

"Several studies have focused on data practices of data scientists; our
research extends the focus on data to ML practitioners, including
engineers, researchers, and academics who build and deploy AI/ML
technologies."

File: papers/niels_bantilan/paper.rst [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+This introductory section provides a brief primer on data validation with pandera,
+providing insights into how its design facilitates code-first schema authoring
+and maintenance. The operating assumption is that this, in turn, gives rise to
+safer and more robust data pipelines.
+
+Why Validate Data?
+++++++++++++++++++
+
+Data validation is the act of falsifying data against a particular set of
+assumptions, expressed as a schema of validation rules. These rules are explicitly
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: Nit-picky but I think data validation is more about validating or verifying than "falsifying." Also, how about using "expectations" or "desiderata" than "assumptions?"

File: papers/niels_bantilan/paper.rst [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+established by the analyst without interference from automated processes, like
+data profiling, and verified at runtime on real-world data.
+
+In machine learning (ML) and statistical analysis use cases, this is critical
+because invalid data, e.g. incorrect types, invalid values, and otherwise
+corrupted data, can pass silently along a data pipeline and propagate those
+errors to various endpoints. These endpoints can be models, analyses, and
+visualizations, which cause adverse ripple-effects to the downstream consumers
+that rely on high-quality data, such as scientific research and business-critical
+applications.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: I like that you mentioned silent failure issue but I think you can also talk about the trust and quality assurance aspect of a bad data problem. Also, I think data validation should be a must for any ML projects regardless of whether a project requires high-quality data. Current wording sounds like high-quality data is required for a subset of ML projects.

File: papers/niels_bantilan/paper.rst [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+
+The analyst can build domain knowledge about the data they are working with
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: I think technically, the mentioned type of domain knowledge that ML practitioners can build is _statistical_ domain knowledge, not the domain knowledge from domain experts, which is often referred to as _domain knowledge_, especially because often ML practitioners are not domain experts. The current wording may sound like we're neglecting the actual domain experts.

File: papers/niels_bantilan/paper.rst [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+by inspecting the data via exploratory data analysis (EDA), automatically
+profiling the data with any number of tools at their disposal, or a combination
+of these two approaches. By building a mental model of how their data looks like
+and envisioning a set of constraints that express what the ideal "clean" dataset
+looks like, the analyst can then encode this understanding as a schema that they
+can use to validate new incoming data.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: I actually think you can motivate data validation more strongly here. As a user of pandera, I find it extremely useful to detect data drift, both for model training and post-deployment monitoring.

File: papers/niels_bantilan/paper.rst [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+part of the development process. As stated in :cite:`niels_bantilan-proc-scipy-2020`, to
+lower the barrier to explicitly writing down schemas for maintaining data quality,
+pandera was created with the following design principles in mind:
+
+1. Expressing schemas as code should feel familiar to pandas users.
+2. Data validation should be compatible with the different workflows and tools in
+   in the data science and ML stack without a lot of setup or configuration.
+3. Defining custom validation rules should be easy.
+4. The validation interface should make the debugging process easier.
+5. Integration with existing code should be as seamless as possible.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:96]: This is great but consider mentioning the reasons behind each bullet point in previous paragraphs (they don't have to be long) because as a reader, I am not finding why you came up with these principles. As a user, I found pandera really easy to use (great accessibility) and this is a huge appeal and a big selling point (which could be reasons behind points 1 and 3).

I am seeing that these reasons are elaborated in the following parts of the manuscript, which is great but I still think you can mention briefly before the list, especially because we have the "origins" section. Just a suggestion.

File: papers/niels_bantilan/paper.rst [hunk 8]
```diff
@@ -0,0 +1,566 @@
...
+
+Pandera's Programming Model
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:101]: Before we start this section, I think it might be nice to have some literature review such as whether there have been any packages like pandera and where they fall short.

File: papers/niels_bantilan/paper.rst [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+           float, pa.Check.in_range(0.0, 1.0)
+       ),
+   })
+
+In the example above, we expect our data to have three columns that have
+specific names, data types, and data value constraints. By reading the code
+the analyst themselves or their collaborators can immediately see what the
+minimum requirements are for valid data.
+
+Pandera emphasizes code-first schema authoring and maintenance, which lowers
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: What would be the opposite of code-first schema?

File: papers/niels_bantilan/paper.rst [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+
+   import pandera as pa
+
+   schema = pa.DataFrameSchema({
+       "column1": pa.Column(int, pa.Check.gt(0)),
+       "column2": pa.Column(str, pa.Check.isin([*"ABC"])),
+       "column3": pa.Column(
+           float, pa.Check.in_range(0.0, 1.0)
+       ),
+   })
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: It might be better to include argument names because readers may not be familiar with the package yet.

File: papers/niels_bantilan/paper.rst [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+
+In the example above, we expect our data to have three columns that have
+specific names, data types, and data value constraints. By reading the code
+the analyst themselves or their collaborators can immediately see what the
+minimum requirements are for valid data.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: Here, you can mention the example above and elaborate what `pa.Check.gt(0)` means (because other things should be familiar to ordinary python users).

File: papers/niels_bantilan/paper.rst [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+and maintenance. The operating assumption is that this, in turn, gives rise to
+safer and more robust data pipelines.
+
+Why Validate Data?
+++++++++++++++++++
+
+Data validation is the act of falsifying data against a particular set of
+assumptions, expressed as a schema of validation rules. These rules are explicitly
+established by the analyst without interference from automated processes, like
+data profiling, and verified at runtime on real-world data.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: How is data profiling different from EDA and data validation?

File: papers/niels_bantilan/paper.rst [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+being out of range:
+
+.. code-block:: python
+
+   invalid_data = pd.DataFrame({
+       "column1": [1, -1, 3],
+       "column2": ["A", "B", "D"],
+       "column3": [0.2, 0.41, 100.0],
+   })
+   # try to validate as many properties as possible
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: When `lazy=True`, based on my experience w/ pandera, all erroneous cases are captured, correct? "as many properties as possible" sounds like we don't know there's no such guarantee.

File: papers/niels_bantilan/paper.rst [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+   Data quality remains a core concern for practitioners in machine learning,
+   data science, and data engineering, and many specialized packages have emerged
+   to fulfill the need of validating and monitoring data and models. However, as
+   the open source community creates new data processing frameworks - notably,
+   new highly performant entrants such as Polars - existing data quality frameworks
+   need to catch up to support them, and in some cases, the Python community
+   more broadly creates new data validation libraries for these new data frameworks.
+   This paper outlines pandera's journey from being a pandas-only data validation
+   framework to one that is extensible to other non-pandas-compliant dataframe-like
+   libraries.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: It's uncommon to have citations in the abstract but I think since you previously published a paper about pandera before, you might want to mention the previous paper to avoid confusion.

You also mention scope change and various technical and non-technical challenges at the end of the manuscript and I think you can elaborate a bit more here about those because "journey" sounds somewhat vague.

File: papers/niels_bantilan/paper.rst [hunk 15]
```diff
@@ -0,0 +1,566 @@
...
+the ``.failure_cases`` attribute. This is reported at the most granular level
+so that the analyst can quickly understand what's wrong with their data.
+
+
+Evolution
+---------
+
+After its first set of releases, pandera continued to improve with bug fixes,
+feature enhancements, and documentation improvements. This section highlights
+four major events in pandera's development.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: Readers would want to see what these 4 major events are in this paragraph.

File: papers/niels_bantilan/paper.rst [hunk 16]
```diff
@@ -0,0 +1,566 @@
...
+
+Documentation is one of the most critical pieces to any software project. Even
+if the underlying code is well-written, performant, and useful, ultimately if the
+documentation is unclear or otherwise difficult to read and navigate, the software
+itself will be inaccessible to end users.
+
+The first set of major contributions came with the help of Nigel Markey, who
+helped considerably in documentation efforts, making pandera easy to learn and adopt.
+This helped pandera to become part of pyOpenSci :cite:`pyopensci`, which helped further
+improve its quality and usability through further review and refinement.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: This sounds vague. Can you elaborate? e.g., "easy to learn and adopt" -> in what way specifically?

File: papers/niels_bantilan/paper.rst [hunk 17]
```diff
@@ -0,0 +1,566 @@
...
+
+Class-based API
++++++++++++++++
+
+The second major improvement in pandera was contributed by Jean-Francois Zinque,
+who implemented the class-based syntax that's more akin to Python dataclasses
+and the pydantic library :cite:`pydantic`. This modernized pandera to use syntax that was familiar
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:260]: It'd be nice to have some feature comparison between pandera and pydantic.

File: papers/niels_bantilan/paper.rst [hunk 18]
```diff
@@ -0,0 +1,566 @@
...
+
+
+Data Synthesis Strategies
++++++++++++++++++++++++++
+
+The third major improvement was adding support for data synthesis strategies
+using the hypothesis library :cite:`MacIver2019Hypothesis`. This expanded pandera's scope from a data
+validation library to a “data testing” toolkit by allowing the analyst to easily
+create mock data for testing not only real data, but the functions that
+produce/clean/transform the data.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: Hypothesis testing is already mentioned in your first paper. How is this different from it?

File: papers/niels_bantilan/paper.rst [hunk 19]
```diff
@@ -0,0 +1,566 @@
...
+   # This will fail on the output check
+   @given(WrongInput.strategy(size=3))
+   def test_fn_wrong_input(input_data)
+       with pytest.raises(pa.SchemaError):
+           fn(input_data)
+
+Hypothesis handles generating valid data under the pandera schema's constraints,
+which relieves the developer from manually hand-crafting dataframes and allows
+unit tests to catch edge cases that would not otherwise be caught by the
+hand-crafted test cases.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: Can this be covered by pytest suite (as a part of test coverage)? I am curious to know how pandera's data validation checks work seamless w/ existing pytest tests.

File: papers/niels_bantilan/paper.rst [hunk 20]
```diff
@@ -0,0 +1,566 @@
...
+
+
+Evolution
+---------
+
+After its first set of releases, pandera continued to improve with bug fixes,
+feature enhancements, and documentation improvements. This section highlights
+four major events in pandera's development.
+
+Documentation Improvements
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:241]: About the order of the 4 points in the manuscript, is it based on chronological order of development or based on the importance of the features?

File: papers/niels_bantilan/paper.rst [hunk 21]
```diff
@@ -0,0 +1,566 @@
...
+hand-crafted test cases.
+
+
+Pandera Type System
++++++++++++++++++++
+
+Finally, the fourth major improvement was contributed by Jean-Francois Zinque,
+who implemented pandera's type system, which provides a consistent interface for
+defining semantic and logical types not only for pandas, but also potentially for
+other dataframe libraries.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: What are other dataframe libraries? Did you also mean to include "custom" types?

File: papers/niels_bantilan/paper.rst [hunk 22]
```diff
@@ -0,0 +1,566 @@
...
+growing community of pandera users also began to expand pandera's scope to
+support pandas-compliant data frameworks such as GeoPandas :cite:`kelsey_jordahl_2020_3946761`,
+Dask :cite:`matthew_rocklin-proc-scipy-2015`, Modin :cite:`petersohn2020scalable`,
+and Pyspark Pandas :cite:`pyspark-pandas` (formerly Koalas).
+
+Design Weaknesses
++++++++++++++++++
+
+As requests for other dataframe-like libraries increased in frequency, it became clear that
+pandera in its existing state was not well-suited for extension beyond Pandas objects.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: I actually think this is a very interesting point in relation with the 5 principles in the first part of the paper. Do you think this shifting of scope made you compromise any of those principles?

File: papers/niels_bantilan/paper.rst [hunk 23]
```diff
@@ -0,0 +1,566 @@
...
+
+
+Expanding Scope
+---------------
+
+After gaining traction over the years, the author, the contributors, and the
+growing community of pandera users also began to expand pandera's scope to
+support pandas-compliant data frameworks such as GeoPandas :cite:`kelsey_jordahl_2020_3946761`,
+Dask :cite:`matthew_rocklin-proc-scipy-2015`, Modin :cite:`petersohn2020scalable`,
+and Pyspark Pandas :cite:`pyspark-pandas` (formerly Koalas).
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: It might be nice to have a connecting sentence that makes transition to design strengths and weakness more natural.

File: papers/niels_bantilan/paper.rst [hunk 24]
```diff
@@ -0,0 +1,566 @@
...
+
+Design Weaknesses
++++++++++++++++++
+
+As requests for other dataframe-like libraries increased in frequency, it became clear that
+pandera in its existing state was not well-suited for extension beyond Pandas objects.
+The fundamental design flaw in pandera's internals was that the schema specification
+and validation engine were interleaved through out the code base. This presented the
+following challenges for supporting non-pandas dataframe libraries:
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:406]: Since this paper is about the journey of evolution, it might be interesting to talk about how you discovered the weaknesses. Did you run tests? Is it just accumulated knowledge over time? Reported by users? etc.

File: papers/niels_bantilan/paper.rst [hunk 25]
```diff
@@ -0,0 +1,566 @@
...
+
+Rewriting Pandera Internals
+---------------------------
+
+For practical purposes, the first set of DataFrame libraries supported by pandera
+were pandas-compliant frameworks such as GeoPandas, Modin, Dask, and Koalas (now ``pyspark.pandas``).
+Even though these libraries do deviate somewhat from the pandas API, they were
+close enough such that the parts of the pandas API that pandera leveraged were just a
+subset of the full API. Therefore, supporting these additional libraries required
+only a few code changes. This approach was the path to least resistance for making data
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: "only a few code changes" -> Would be nice to see an example

File: papers/niels_bantilan/paper.rst [hunk 26]
```diff
@@ -0,0 +1,566 @@
...
+   import sloth as sl
+
+   from pandera.api.extensions import register_builtin_check
+
+   @register_builtin_check(aliases=["eq"], error="equal_to({value})")
+   def equal_to(data: sl.Series, value: Any) -> sl.Series:
+       return data.equals(value)
+
+
+Organizational and Development Challenges
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:526]: I actually really appreciate that you mention these challenges!

File: papers/niels_bantilan/paper.rst [hunk 27]
```diff
@@ -0,0 +1,566 @@
...
+
+Retrospective
++++++++++++++
+
+With all of these challenges in mind, the internals rewrite was completed in
+pull request 913 :cite:`pandera-pr-913` on January 24th, 2023 and the follow-up
+pull request 1109 :cite:`pandera-pr-1109` on March 13th, 2023. A few factors
+facilitated the rewrite itself and also reduced the risk of regressions:
+
+- **Unit tests**: A comprehensive unit test suite caught many issues, but not all of them.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst]: What was particularly more difficult to find?

File: papers/niels_bantilan/paper.rst [hunk 28]
```diff
@@ -0,0 +1,566 @@
...
+  e.g. by using native python data structures like lists and dictionaries instead
+  of pandas DataFrames to report failure cases.
+- **Decoupling metadata from data**: Distinguish between DataFrame metadata schema
+  errors (e.g. missing columns) and data value errors (e.g. out-of-range values).
+- **Investing in governance and community**: Invest more in governance and formalize
+  contributor and community RFC processes sooner to help with design and feature
+  enhancement efforts.
+
+
+Updated Design Principles
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:588]: Oh I actually really love that the paper ends with updated principles. Really nice structure :)

File: papers/niels_bantilan/paper.rst [hunk 29]
```diff
@@ -0,0 +1,566 @@
...
+----------
+
+Pandera has evolved from a pandas-specific data validation library to a
+comprehensive toolkit that provides a standard schema interface for easily extending
+and supporting validation backends for arbitrary statistical data containers.
+This paper provides an overview of data validation and testing, focusing on
+pandera's core programming model and its extended functionality to support
+property-based testing. This paper also provides a useful case study of the
+technical and organizational challenges associated with expanding the scope of a
+library beyond its original boundaries.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:616]: Would be nice if this is mentioned in abstract a bit.

File: papers/niels_bantilan/paper.rst [hunk 30]
```diff
@@ -0,0 +1,566 @@
...
+
+The author's hope is that, by highlighting the technical and organizational
+dimensions of this evolution, that other open source authors and maintainers can
+learn and avoid some of the pitfalls encountered during the internals rewrite
+that now enables pandera to support a whole suite of statistical data containers
+moving forward.
```
Comment (hongsupshin) [inline papers/niels_bantilan/paper.rst:622]: Love this motivation 👏🏼

**Review by hongsupshin — COMMENTED** [review]
Hey Niels, great paper! I've been using pandera myself, so first of all, I am really glad that I got to review this paper :) 

Most of my comments are suggestions to make your paper more clear and accessible. Hope they help.

File: papers/niels_bantilan/paper.rst (hunk 1 above)
Comment (hongsupshin) [reply papers/niels_bantilan/paper.rst:52]: Now that I have read the whole manuscript, I can see why you split the paper into origins and evolution but I think there should be an introductory paragraph before the origins to talk a bit about the structure of the paper, which might improve readability.

**Comment by cosmicBboy** [comment]
Thanks for the great feedback @hongsupshin ! will work on incorporating this over the next few weeks

**Comment by cbcunc** [comment]
@cosmicBboy We are beyond the final review deadline. How is it going addressing the review comments? Would you please mark the review comments as resolved as you add commits to address them? Thank you.

**Comment by cosmicBboy** [comment]
hi @cbcunc apologies for the delay! I'll push up changes later tonight and mark them as resolved

File: papers/niels_bantilan/paper.rst (hunk 17 above)
Comment (cosmicBboy) [reply papers/niels_bantilan/paper.rst:260]: skipping this, but a good topic for a blog post!

File: papers/niels_bantilan/paper.rst (hunk 19 above)
Comment (cosmicBboy) [reply papers/niels_bantilan/paper.rst]: pandera schemas are just fancy assertions on dataframes, so pytest is orthogonal to pandera validation checks.

File: papers/niels_bantilan/paper.rst (hunk 22 above)
Comment (cosmicBboy) [reply papers/niels_bantilan/paper.rst]: > Do you think this shifting of scope made you compromise any of those principles?

A little bit, hence the amended principles :)

File: papers/niels_bantilan/paper.rst (hunk 3 above)
Comment (cosmicBboy) [reply papers/niels_bantilan/paper.rst]: kept "assumptions"

**Comment by cosmicBboy** [comment]
added the updates and resolved comments!

**Comment by hongsupshin** [comment]
Looks good to me :) Thanks for resolving the comments @cosmicBboy !

**Comment by cbcunc** [comment]
@scoobies mark pending comment

**Comment by cbcunc** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.25080/Majora-92bf1922-00a is OK
- 10.25080/Majora-342d178e-010 is OK
- 10.21105/joss.01891 is OK
- 10.5281/zenodo.3946761 is OK

MISSING DOIs

- 10.25080/majora-7b98e3ed-013 may be a valid DOI for title:  Dask: Parallel Computation with Blocked algorithms and Task Scheduling 
- 10.14778/3407790.3407807 may be a valid DOI for title: Towards Scalable Dataframe Systems

INVALID DOIs

- None
```

**Comment by cbcunc** [comment]
@cosmicBboy Would you please address the missing DOIs above and then your paper will be ready for final editor review. Also, you do not need to merge the parent branch into yours (in fact, it is preferable that you don't as per the author instructions, although there was no harm done by your latest merge commit).

**Comment by cosmicBboy** [comment]
just added the missing DOIs

**Comment by cbcunc** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.25080/Majora-92bf1922-00a is OK
- 10.25080/Majora-342d178e-010 is OK
- 10.21105/joss.01891 is OK
- 10.5281/zenodo.3946761 is OK
- 10.25080/Majora-7b98e3ed-013 is OK

MISSING DOIs

- None

INVALID DOIs

-  https://doi.org/10.48550/arXiv.2001.00888  is INVALID because of 'https://doi.org/' prefix
```

**Comment by cbcunc** [comment]
@cosmicBboy Fixing the missing DOIs add an invalid DOI. Please see above. It just needs a naked DOI instead of one preceded by https://doi.org/. Would you please fix this? We are having our final review today and this is one of the few papers not in the ready state.

**Comment by cosmicBboy** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.25080/Majora-92bf1922-00a is OK
- 10.25080/Majora-342d178e-010 is OK
- 10.21105/joss.01891 is OK
- 10.5281/zenodo.3946761 is OK
- 10.25080/Majora-7b98e3ed-013 is OK
- 10.48550/arXiv.2001.00888 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by cosmicBboy** [comment]
@cbcunc fixed the offending DOI, do things look good now?

**Comment by cbcunc** [comment]
@scoobies mark ready

**Review by deniederhut — APPROVED** [review]
Thanks @hongsupshin for helping us get this paper ready for publication! And thanks @cosmicBboy for 

source-1012164550
joining us at SciPy this year!
